src/visualizers/VisualizerGrid.ts

```diff
@@ -0,0 +1,1041 @@
+            required: false,
+            visibleDependency: index > 0 ? `property${index - 1}` : '',
+            visiblePredicate: (d: Property) => d !== Property.None,
+        },
+        [`prop${index}Vis`]: {
+            value: prop.visualization,
+            from: PropertyVisualization,
+            displayName: 'Display',
+            required: false,
+            visibleDependency: `property${index}`,
+            visiblePredicate: (d: Property) => d !== Property.None,
+        },
+        [`prop${index}Color`]: {
+            value: prop.color,
+            forceType: 'color',
+            displayName: 'Color',
+            required: false,
+            visibleDependency: `property${index}`,
+            visiblePredicate: (d: Property) => d !== Property.None,
+        },
+        [`prop${index}Aux`]: {
+            value: prop.aux,
+            displayName: (d: Property) => propertyAuxName[Property[d]] || '',
+            required: false,
+            visibleDependency: `property${index}`,
+            visiblePredicate: (d: Property) => Property[d] in propertyAuxName,
+        },
+    }
+}
+
+/*
+ * The following functions are used to check properties.
+ */
+
+function isPrime(factors: Factorization): boolean {
+    if (factors === null) throw new Error('Internal error in Grid')
+    if (factors.length == 0) return false // 1 is not prime
+    const [factor, power] = factors[0]
+    switch (factors.length) {
+        case 1:
+            return factor > 1n && power === 1n
+        case 2:
+            return factor === -1n && factors[1][1] === 1n // negative of prime
+    }
+    return false // two or more prime factors
+}
+
+// Adapted from Geeks for Geeks:
+// https://www.geeksforgeeks.org/deficient-number/
+function getSumOfProperDivisors(num: bigint): bigint {
+    if (num === 0n) return 1n // conventional value to make 0 abundant.
+    // returns the sum of divisors of the absolute value
+    if (num < 0n) num = -num
+    let sumOfDivisors = 1n // 1 always divisor, leave out num itself
+    let i, partner
+    // Note that this loop will break beyond square root of n
+    for (i = 2n, partner = num / i; i <= partner; i++, partner = num / i) {
+        if (num % i === 0n) {
+            sumOfDivisors += i
+            if (partner > i) sumOfDivisors += partner
+        }
+    }
+    return sumOfDivisors
+}
+
+function isSumOfTwoSquares(factors: Factorization): boolean {
+    if (factors === null) throw new Error('Internal error in Grid')
+    let legendre = 1
+    for (let i = 0; i < factors.length; i++) {
+        const factor = factors[i]
+        const prime = factor[0]
+        const exponent = factor[1]
+        if (prime === -1n) {
+            // negative numbers are never sums of squares
+            return false
+        }
+        // otherwise update legendre for every prime 3 mod 4
+        if (prime % 4n === 3n && exponent % 2n === 1n) {
+            legendre = -1 * legendre
+        }
+    }
+    if (legendre === 1) {
```

Comment:
I will simplify this.

Reply:
Done in https://github.com/numberscope/frontscope/pull/230/commits/4e85af069d6f9c3c915403272a00a8cef809b2ff.

Reply:
Agreed. Resolving.